Plugin fails with post 1.3.0 Redmine: Code included to fix (backward compatible)

kubitron opened the pull request:
There are three problems that have arisen with running redmine_git_hosting on the 1.3-stable branch of Redmine (DO NOT PULL THE TRUNK -- IT IS HAS A PARTIALLY IMPLEMENTED FEATURE THAT WILL REALLY BREAK THIS PLUGIN AS IT CURRENTLY EXISTS):

1) The developers have removed the TabularFormBuilder, thus breaking several views
2) The developers have deprecated (for removal in 1.5) labelled_tabular_form_builder, which will break several views.
3) The developers have removed the generic :controller/:action/:id route at the bottom of routes.rb, thereby breaking the edit/delete functions for gitolite keys.

This pull request fixes these problems (and cleans up the gitolite-key editing functionality in the process). Note that it is referenced off my master branch, which provides a bunch of other benefits in stability (includes the Resilience patches discussed in #124). If there is sufficient interest to somehow factor it back to the current redmine_git_hosting master, I will consider requests... Incidentally, I believe that this fix _should_ be backward compatible with earlier versions of Redmine (that is my intention!).  Let me know if it is not true (at least works with 1.2.2, and should work with at minimum 1.0, but...)

--EDIT 1/12/2012 --
With latest commit ([00aa93d](https://github.com/kubitron/redmine_git_hosting/commit/00aa93d81b0b20d6f7c69d8a509d0df9c9014454)), now export user keys to administrator off the user/edit page, also fixed slight security flaw.

-- Edit 1/16/2012 --
One additional commit ([0740c23](https://github.com/kubitron/redmine_git_hosting/commit/0740c239a4f08595f5686dcb93a808f5c3093612)) will resync the hooks parameters on a fetch_changesets, making sure that the update keys don't get out of sync. This is really a part of the resilience branch (which underlies this set of commits).

It is very important to execute a db:migrate_plugins after installing this, since the other commits on which it depends have introduced new plugin options.  Thus, make sure to do:

```
rake db:migrate_plugins RAILS_ENV=production
```

Please get back to me with any bugs.

kubitron committed:
Added hooks, policy and rakefile to install selinux policy for this plugin

kubitron committed:
Performance fix for redmine_git_hosting.

This prevents a rediculous number of calls to update_repositories
caused by the fact that redmine saves to the repository model for
every commit that it examines.  This patch also prevents multiple
calls to update_repositories when adding groups to the membership
of a project.

You need to migrate plugins to take advantage of new indices:

	rake db:migrate_plugins

kubitron committed:
More performance improvements.

This patch should remove all remaining need for recursive calls to update_repositories.
Will log such calls as errors in upcoming major code fix...

kubitron committed:
Rewrite of primary gitolite interface for better resilience.

The primary "update_repositories" function is now largely self-correcting
against a variety of errors seen in the field.  Also added a bunch of
error checking on functions called-out to the shell.  Among other things,
this code attempts to reconnect the gitolite-admin public key when it
gets deleted (can happen).

Regular use of 'sys/fetch_changesets' will recorrect any errors in the
public key directory and gitolite.conf file. For example, orphan keys
that happen to be left behind in the public key directory will be removed,
missing keys will be added, etc.

One new piece of functionality is the 'recycle_bin' which is used to store
deleted repositories immediately after they are deleted (allowing them
to be recovered for up to 'gitRecycleExpireTime' hours after deletion).

You need to migrate plugins because there are new plugin settings:

    rake db:migrate_plugins

kubitron committed:
Missed a performance case/source of recursion complaints in the log.

This is not a major issue (or even correctness issue), but will cause
complaints of recursion in the log.

kubitron committed:
Update of selinux policy for latest version of selinux package. Also
change rakefile to install binary policy (redmine_git.pp) instead of
compiling source policy (redmine_git.te) before every
installation. New rake task added to recompile policy if necessary:

    rake selinux:redmine_git_hosting:build_policy

kubitron committed:
Added code to correct for changes in the Redmine 1.3->1.3.1 transition.

First, altered views to correct for the fact that (1) TabularFormBuilder
was removed and (2) labelled_tabular_form_for was deprecated.  To allow
this new code to work with pre 1.3.1 versions of Redmine, added an
app/helpers/application_helper.rb which defines the new accepted
helpers if they are not already defined.

Second, made the gitolite_public_keys controller a bit more RESTful in
response to the removal of the generic :controller/:method/:id route.
Note that without these changes, get weird failures around the /my/account
sidebar.  Now, everything happens on the /my/account page: Does not use
separate screen for key edit, and removed the confusion "active" checkbox
on edit (given recent fixes, this is only a weird way to delete key).  Finally,
this code asks for confirmation before deleting a key.  Note that account.rhtml
updated slightly (has password icon now near "change password").

kubitron commented:
Sorry, all. I accidentally pushed a bug into the second-to-last commit. Fixed through history editing (bad idea, I know, trying to keep packaging simple, should these commits ever be merged into the main branch.

Just kill of last two commits and repull.

Would love comments from others on this latest commit. This commit will be the last in this resilience branch (promise), assuming that it works for people.

Any of the old lurkers still out there? bakerjonas? simonsd? Willing to try this new commit?

simonsd commented:
Sorry, running this in beta for my company's internal infra right now.
Don't really feel like breaking it atm.

I am however planning on setting up a new vm tomorrow for testing purposes, might test the patch there.
For now I'll stick to running 1.2.1, might try out migration to 1.3.x later on.
I'll keep you informed :)

kubitron commented:
Actually, this fix is only compatible with 1.3x -- doesn't need to use it.  Should work find with 1.2.1!

(And have other benefits!).  I'm more interested in the previous commits anyway... They continue the fixes that you liked on my resilience pull request (#124).

kubitron committed:
Continuing rewrite of redmine_git_hosting to rationalize path handling.

It is important to remember to migrate_plugin settings, since there
are new settings and one setting has changed a bit (i.e. httpServer
should nolonger include a rails_root path).  Thus, make sure to:

    rake db:migrate_plugins

This commit includes a number of changes.  First and foremost, the
update_repositories() function now handles deletes and repository
moves as well as all other changes.  Consequently, :RESYNC_ALL will be
able to recover from a variety of problems involving movement of
repositories resulting from project parent changes.  Note that changes
in the parent for a subtree of projects works much better than before.

Second, this commit adds new settings which allow (1) redmine-managed
repositories to be focused in a specific subdirectory of the gitolite
repository, (2) allow repositories to be named either hierarchically
(default) or concentrated in a single directory (flat) independent of
project parents, and (3) an extra path parameter which can be added to
the http URL for smart http access of repositories.

Third, added the "access" box at the top of the settings page which
shows explicitly how settings affect access parameters.

Random improvements include (1) rewrite of routes for smart HTTP to
better handle changes in parentage (and avoid need to change routes as
project parents change); (2) changes in settings now immediately
trigger resulting changes in state (before, the settings cache got in
the way); (3) all of the path functions (repository positions, ssh
access, http access) have been concentrated into a small number of
functions at the top of git_hosting.rb, rather than spread throughout
the code; (4) continuing the changes started with the original
performance fixes, cleaned up observer behavior triggered by changes
in project membership and settings.

kubitron committed:
Add user-edit screen for public_keys and fix security flaw.

This commit refactors the gitolite-public-key view to (1) remove
MVC violations (i.e. move the DB access into respective controllers,
(2) separate the view/edit functionality into a separate partial so
that (3) it can be called from the user edit screen.  As a result,
the administrator can now look at and edit user's keys (which I have
found is necessary for some clueless users).  The previous commit
would allow someone to view other public keys (but not edit them!); this
commit fixes this flaw.

Also, updated the style of the display to better reflect the 1.3+
style of redmine.  This code is still backward compatible with
earlier versions of redmine.  Also, moved .rhtml => .html.erb in
preparation for Rails 3.0 (rest of Redmine alreday updated)

kubitron committed:
Repair hook parameters for all projects during :resync_all.

kubitron committed:
Allow repair hook functionality to work with older versions of sudo.

kubitron committed:
Added support for ssh port specification in gitServer. Also clarity fix for access box.

kubitron committed:
Fix bugs in error path for repository mirror push. Correct status is now reported. Removed extra cruft in error reporting.

kubitron committed:
Encode username in HTTP URL bar so that resulting URLS are valid for
use with smart HTTP, even when usernames include illegal characters
such as '@'. For example, usernames such as "John@Doe" => "John%40Doe"
in the URL box.

kubitron committed:
Fixes for compatibility with Redmine 1.1

gcymbalski committed:
fixes some stray semicolons that broke ruby 1.9

kubitron committed:
Merge pull request #15 from gcymbalski/chiliproject-and-1.9

Fixed some stray colons that broke ChiliProject 3.0 on Ruby 1.9.2-p290.
According to gcymbalski, ChiliProject is 3.0 is now functional.

kubitron committed:
Make temp and scripts directory configurable. Also, add default
repository access methods for new repositories (i.e. for daemon, HTTP,
and CIA notification). Finally, added more rake tasks to ease
maintenance (in the :redmine_git_hosting namespace). Update README
with information about these new options and with with new
compatibility information.

Since there are 5 new settings, make sure to migrate plugins again:

    rake db:migrate_plugins RAILS_ENV=xxx

kubitron committed:
Update the installation process to explicitly populate the scripts
directory, thereby (hopefully) avoiding confusion about this
directory. Give options for making scripts read-only if desired. Small
updates to settings logic to better support this installation flow.

kubitron committed:
Make sure that log messages flushed to log for rake tasks. Also, mark
tasks in gitolite.rake as deprecated.

kubitron committed:
Recover gracefully from missing git_repository_extra structure. Did
this by changing the access mechanism to automatically construct a
git_repository_extra structure if it is missing.

kubitron committed:
Fixed git-daemon support. Now, the "daemon" key is properly added and
removed as project and repository permissions change. Also, when
projects are not public, the git-daemon selection box on the
repository page is disabled (to make it clearer that non-public
projects cannot be exported by the git daemon).

kubitron committed:
Better error recovery from failure to pull administrative repo.

Several changes here: (1) if failed to merge changes, try deleting and
recloning before doing something more drastic.  (2) Try to fixup admin
key by using direct cloning of admin repo and pushing of changes via
gl-admin-push; fallback to gl-setup only if gl-admin-push doesn't
exist.  (3) When reinstalling admin key, try to keep admin key name
from conf file (if exists) or keydir (if there is a matching key to
remine's key).  (4) Finally make sure to remove any keys from keydir
that match redmine's key before reinstalling key -- to avoid name
conflicts on access.

kubitron committed:
Fixed slight bug: missed case in which removing user from membership
of project should have triggered removal of user's keys from the
keydir.

kubitron committed:
Fixed post-receive mirror push so that multiple mirrors will be updated.
This problem likely introduced by earlier fix to mirror error reporting.

kubitron committed:
Fixed language tags on localizations.

kubitron committed:
Extra protection when defining backward-compatible versions of
labelled_form_for and labelled_remote_form_for: do not do so if
lib/tabular_form_builder.rb undefined!

kubitron committed:
Incremented redmine_git_hosting version from 0.4.3x => 0.4.4x.

Changes in this release:

1) Compatibility with Redmine 1.1-stable
2) Compatibility with Ruby 1.9 (really just some syntax fixes)
3) Better configurability, including ability to set /bin and /tmp paths.
   Better scripts to aid in installation.  Updated instructions in README.
4) Variety of bug fixes: mirrors work better, git-daemon support, better
   error recovery from lost gitolite admin key, handling of '@' in username
   for smart-http access.

kubitron committed:
git_httpd_controller fix to deal with null buf (which is possible).
Grabbed fix from Eric Bishop (who received via email by Davil Vallejo).

kubitron committed:
Added more explicit instructions for configuring gitolite and generating
gitolite administrative key.

kubitron committed:
Further Compatibility fixes for Redmine 1.1/backward compatibility

1) Allow the csrf_meta_tag() in popup.erb to fail gracefully, since
version skew in the Rails installation on some Debian platforms causes
this operation to fail (complaining about a missing String#html_safe).

2) Also in popup.erb, if helper function "javascript_heads" not
defined (it isn't in Redmine 1.1), simply use "javascript_include_tag"
to get javascript included.

kubitron committed:
Fixed weird race condition with cron cleanup of /tmp (such as tmpwatch).

Now, clone_or_pull_gitolite_admin() will refuse to merge in existing
changes (which could involve deletes from tmpwatch).  Instead, it will
recognize the unmerged changes and simply delete and reclone.

Note that such unmerged changes will not exist unless something bad
has happened.  Cleanup may require executing a RESYNC_ALL operation
(i.e. fetch_changesets).

kubitron committed:
Incremented redmine_git_hosting version from 0.4.4x => 0.4.5x

This is a bugfix release.  Changes in this release:

1) Fixed missed case for Redmine 1.1 compatibility (mirror functionality)
2) Fixed bad interaction between cron cleanup of /tmp and access to
   gitolite-admin repository in /tmp.  Behavior could cause user
   keys to appear to be deleted, even though they remain in the redmine
   database.

xdissent committed:
Initial post-receive urls implementation

xdissent committed:
Add migrations for post-receive urls

xdissent committed:
Fix typo in post-receive urls controller

xdissent committed:
Typo in repository partial

xdissent committed:
Ajaxify post-receive url settings

xdissent committed:
Refactor extracting repo info for hooks

xdissent committed:
Fill out post-receive url payload

xdissent committed:
Only count response code 200 as success in post-receive hook

xdissent committed:
POST to post-receive URLs and add en translations

xdissent committed:
Adding post-receive URL modes

xdissent committed:
Add GET style hook and missing translation

kubitron committed:
Fixup key identifiers in (rare) case in which login name changes.

kubitron committed:
Make sure that reset_identifier works with RAILS 3

kubitron committed:
Fix bug with display of gitolite_public_keys when failing User validations.

kubitron committed:
Properly handle the archiving of projects.

When a project is archived, we remove its keys from gitolite.conf (and
possibly from the keydir if this project is the only one that uses
these keys).  We also remove the git_daemon access and mark the gitolite.conf
entry with a dummy key: "redmine_archived_project".

We leave the repository in the gitolite repo.

kubitron committed:
Fixed two cases where user delete did not update gitolite properly:

1) Delete keys from gitolite.conf when removing users from project ACL.

Not sure how this case never noticed before.  Bug happens because
user delete => update_repositories(:delete), which didn't go through
and rationalize the gitolite.conf.  Removed optimization to fix.

2) When deleting user, execute use :delete flag to remove keys keydir.

kubitron committed:
Merge branch 'post-receive'

Conflicts:
	init.rb

This is a merge of suggested changes from Greg Thornton
<gthorton@cjadvertising.com> to support invoking post-update URLs.
Conflict was from the fact that I moved modifications of Project into
project_patch.rb (since there are now a number of patches to
project.rb now).

kubitron committed:
Added new parameter 'gitForceHooksUpdate'. When true, will replace
existing hooks if they don't match our hooks. Default value == true.

kubitron committed:
Some fixes to previous merge

Fix syntax error in gitolite_hoooks_controller.rb, turn payload into json.
Addition of changes to missing non-english locals.

kubitron committed:
Remove control characters from key text.

kubitron committed:
Updated ModalBox to latest version.

kubitron committed:
More flexible repository mirror support.

Now, offer ability to push upstream without full mirroring
(i.e. without the '--mirror' argument).  Added ability to force
updates or allow fast forward updates.  Further, can choose to
push references with an explicit reference specification (supports
full syntax, allowing one to push to a branch with a different remote
name than the local name.

kubitron committed:
Rearrange dependencies to prevent failure in testing.

kubitron committed:
Added the ability to use an alternate gitolite configuration file.

New parameters gitConfigFile and gitConfigHasAdminKey set this
feature.  The gitConfigFile is relative to the conf/ directory.  The
gitConfigHasAdminKey parameter states whether or not the plugin will
maintain the administrative key in the gitConfigFile (forced to true
for default config file of gitolite.conf.

kubitron committed:
Fixed bug in post-receive hook which prevented multi-ref notification.

This was due to a bug in ruby 1.8.7's Net::HTTP::Post.set_form_data.

kubitron committed:
Remove dispatcher from init file, since this nolonger exists in Rails 3.0.

Reorganized patches to put all of the patch-specific includes
and patching operations into the individual patch files.  This leaves
only a need for explicitly loading each of the patch files, which
is now done with a require_dependency inside a loop...

kubitron committed:
Fix slight errors on 'my_account' page

kubitron committed:
Removed validation of repository mirror url, but enclose url in quotes
for actual command (to prevent shell expansion of bad characters).

kubitron committed:
Fixed indentation => 4 spaces (far more readable)

kubitron committed:
First take on implementation of deployment credentials.

Deployment credentials are constructed on the repository settings page
for a repository and consist of combinations of public keys and
permissions (currently either "R" or "RW+".  The public keys for
deployment credentials are owned by users with 'manager' permissions.
The administrator can modify (add/update/delete) all deployment
credentials.

Also included with this commit is a functioning validation on the
public key display window (this wasn't working originally).

kubitron committed:
Fix formatting on "my account" page to match more recent changes.

kubitron committed:
Validate global uniqueness of public keys.

kubitron committed:
Relax public key validation for old (possibly non-conforming) keys.

kubitron committed:
Better handling of projects with a disabled repository module.

We now treat this case exactly the same as an archived
project/repository.  Before, it was handled inconsistently.  Three
changes:

1) When a repository module is disabled/enabled through settings page
   call update_repositories to update state.  A disabling of the
   repository module will call update_repositories with :archive.

2) Existing repositories will be left alone and marked with a dummy
   redmine key of "redmine_disabled_project".  Reenabling of the
   repository module will restore project and keys to accessibility.

3) When removing redmine keys from a gitolite.conf entry, don't report
   in the log unless removing real readmine keys.  This avoids a weird
   circumstance in which every fetch_changesets operation reports that
   it is deleting redmine keys (caused when repo entry exists but the
   "delete_repositories" flag is disabled at the time that the repo is
   deleted.  This fix is entirely about removing confusing messages
   from the log.

kubitron committed:
Remove excess whitespace before validation.

kubitron committed:
Formatting tweak: bigger key text box.

kubitron committed:
Incremented redmine_git_hosting version from 0.4.5x => 0.4.6x

This is a feature release.  It includes:

     1) Post-Receive URLS
     2) Updated Mirror Functionality
     3) Deployment Credentials
     4) Support for secondary gitolite configuration file

It also includes a series of bug fixes.  See release notes at:
https://github.com/kubitron/redmine_git_hosting/wiki/_history

kubitron committed:
Fixed non-standard condition that tripped-up SQLite3.

kubitron committed:
First take at integration (basically works for 1 repo).

kubitron committed:
Working version of cached_shell_redirector, still with debugging.

kubitron committed:
Mostly working integration with 1.4. Still with debugging.

kubitron committed:
Removed much of the debugging for 1.4 port. Some bug and performance fixes.

kubitron committed:
Moved git_caches.proj_identifier => repo_identifier.
Also, check for repository_identifier uniqueness when migrating.

kubitron committed:
Some optimizations, fixes, cosmetic changes. Update README & version => 0.5.0x.

kubitron committed:
Fixed bug in update_repositories

kubitron committed:
Another bug-- didin't combine credentials correctly

kubitron committed:
Fix in <1.4 options edit.

kubitron committed:
Fixed check for repo.ident uniqueness when no repositories.

kubitron committed:
Fix for user-filter error.

kubitron committed:
Remove references to unneeded model box file: builder.js

kubitron committed:
More updating of compatibility information in README.

kubitron committed:
Add empty sys_helper.rb to deal with Ruby 1.9.3

kubitron committed:
Switch back to using dispatcher in Rails 2 for patches

kubitron committed:
More disciplined loading of patch files (well defined order).

kubitron committed:
Removed constraint on Ruby < 1.9.3 from README

kubitron committed:
Don't index repository identifier if doesn't exist

kubitron committed:
Fix migration/undo migration issues associated with changesets.notified_cia

kubitron committed:
Don't look at repository identifier unless we are in multi-repo mode.

kubitron committed:
Fixed other cases that looked at repo identifiers in single-repo mode

kubitron committed:
Fixed problem with pre-1.4 update of members.

kubitron committed:
PostReceive hook now handle https URLs properly.

kubitron committed:
Incremented redmine_git_hosting version from 0.4.6x => 0.5.0x

This is a feature release which provides compatibility with Redmine 1.4.

It also includes a few bug fixes.  See release notes at:
https://github.com/kubitron/redmine_git_hosting/wiki/Release-Notes

kubitron committed:
Fixed bad case for pre-1.4 Redmine edit of Git repo

kubitron committed:
Fix for Ruby 1.9.x: Handle fact that IO.instance_methods can return either array of strings or symbols depending on version of Ruby.

kubitron committed:
Patches to aid initial migration, suggested by Greg Thornton

kubitron committed:
Fixed case for Redmine < 1.4 with project having no Git repo.

kubitron committed:
Redo migration to deal with PostGreSQL (no rescue)

kubitron committed:
Make hook work with Ruby 1.9.x

kubitron committed:
Incremented redmine_git_hosting version from 0.5.0x => 0.5.1x

This is a bug-fix release to restore compatibility with Redmine < 1.4
and with Ruby 1.9.x.  See release notes at:

    https://github.com/kubitron/redmine_git_hosting/wiki/_history

crocket committed:
Fix an error that occurs while uninstalling from DB

2011072600000_extend_changesets_notified_cia.rb tries to delete
notified_cia column in changesets table, but
2011081700000_move_notified_cia_to_git_cia_notifications.rb deletes
notified_cia column before it does, so an error occurs.

A solution is to let 2011072600000_extend_changesets_notified_cia.rb
remove notified_cia column only if the column exists.

kubitron committed:
Only create initial repo if project successfully created.

kubitron committed:
Fix to prevent new repo creation from claiming existing repos.

tomka committed:
Remove unneeded whitespace in gitolite config comment

kubitron committed:
Merge pull request #98 from tomka/fix/gitolite-config-comment-cleanup

Remove unneeded whitespace in gitolite config comment

kubitron committed:
Fix for quoting with an old sudo.

kubitron committed:
Fix to allow unrelated administrative keys in the config file.

This fix involves the key repair code as well as the configuration file
handling code (to maintain multiple administrative keys).

kubitron committed:
Let repositories/_navigation partial be handled by Redmine.
